### PR TITLE
`Exam mode`: Improve suspicious behavior dashboard 

### PR DESCRIPTION
**Pull request by tobias-lippert**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

We should analyze more data and makes it configurable what the instructor wants to analyze.

### Description

The PR does the following things
- [x] align all cases in the table
- [x] find cases where a student exam has sessions with different ip address and/or browser fingerprint
- [x] allow instructors to find exam sessions  with an IP address outside of a specified range (subnet)
- [x] make it configurable for an instructor what they want to analyze.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 or more  Students
- 1 running exam

1. As a student generate suspicious sessions (join another student exam, one student only joins another exam but isn't registered for the exam/doesn't start the exam, join your exam in another browser etc.)
2. Navigate to the exam checklist
3. Go to suspicious behavior
4. Try all the analysis options and verify that they return the correct result

#### Exam Mode Testing
Covered with the testing steps above.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [exam-session.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5367/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/exam-session.model.ts.html) | 100% | ✅|
| [suspicious-behavior.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5367/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/suspicious-behavior/suspicious-behavior.component.ts.html) | 100% | ✅ |
| [suspicious-sessions-overview.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5367/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/suspicious-behavior/suspicious-sessions-overview/suspicious-sessions-overview.component.ts.html) | 100% | ✅ |
| [suspicious-sessions.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5367/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/suspicious-behavior/suspicious-sessions.service.ts.html) | 100% | ✅ |
| [suspicious-sessions.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5367/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/suspicious-behavior/suspicious-sessions/suspicious-sessions.component.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |Comment|
|------------|--------------:|---------------------:|--------------------------:|
| [SuspiciousExamSessions.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5367/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.exam/SuspiciousExamSessions.html) | 100% | ✅ |
| [SuspiciousSessionReason.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5367/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.exam/SuspiciousSessionReason.html) | 100% | ✅ |
| [SuspiciousSessionsAnalysisOptions.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5367/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.exam/SuspiciousSessionsAnalysisOptions.html) | 100% | ✅|
| [ExamSessionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5367/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.exam/ExamSessionService.html) | 100% | ✅|
| [ExamResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5367/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ExamResource.html) | 89% | ✅ |my changes are covered
### Screenshots

https://github.com/ls1intum/Artemis/assets/84102468/1c53374c-5845-493f-88d0-57c34172091d

### CODE REVIEW

File: src/main/webapp/app/exam/manage/suspicious-behavior/suspicious-sessions/suspicious-sessions.component.ts
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exam/manage/suspicious-behavior/suspicious-sessions/suspicious-sessions.component.ts:6]: Why?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java
Comment (MaximilianAnzinger) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:1325]: Or apparently you already have an Option model.

File: src/main/webapp/i18n/en/exam.json
Comment (laurenzfb) [inline src/main/webapp/i18n/en/exam.json:679]:
```suggestion
                "subnet": "Subnet in CIDR notation, if you enter a IPv4 subnet, only IPv4 addresses will be checked. If you enter a IPv6 subnet, only IPv6 addresses will be checked.",
```

File: src/main/webapp/i18n/en/exam.json
Comment (laurenzfb) [inline src/main/webapp/i18n/en/exam.json:680]:
```suggestion
                "subnetHelp": "The subnet in CIDR notation, e.g. 127.0.0.1/24. If you enter a IPv4 subnet, only IPv4 addresses will be compared. If you enter a IPv6 subnet, only IPv6 addresses will be compared.",
```

File: src/main/webapp/i18n/en/exam.json
Comment (laurenzfb) [inline src/main/webapp/i18n/en/exam.json:689]:
```suggestion
                    "sameIpAddressDifferentStudentExams": "Different student exams with the same IP address",
```

**Review by laurenzfb — CHANGES_REQUESTED** [review]
Overall LGTM but have some small change suggestions

File: src/main/webapp/app/exam/manage/suspicious-behavior/suspicious-sessions/suspicious-sessions.component.ts
Comment (tobias-lippert) [reply src/main/webapp/app/exam/manage/suspicious-behavior/suspicious-sessions/suspicious-sessions.component.ts:6]: this is an attribute selector instead of a component selector and allows using it e.g., like <tr jhi-suspicious-sessions </tr

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java
Comment (tobias-lippert) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:1325]: while I would prefer this, the angular http client doesn't support sending a body with a get request.

**Review by laurenzfb — APPROVED** [review]
LGTM after changes!

**Review by MaximilianAnzinger — DISMISSED** [review]
Code LGTM

**Review by MaximilianAnzinger — COMMENTED** [review]
Manual test on ts2 works fine, but you should fix the styling issues before merging it.

**Review by nityanandaz — COMMENTED** [review]
Manual test on TS2 with some UI bugs to sort out :)

**Review by MaximilianAnzinger — APPROVED** [review]
Reapprove code
Manual: Styling looks now better 👍

**Review by laurenzfb — APPROVED** [review]
reapprove

**Pull request merged**
